### PR TITLE
Update API

### PR DESCRIPTION
**Pull request by Sahl-A**
#### The following updates are done:
- updated the REST client to include variables + added authorization headers
- changed ready column to true whenever created column is set to true as long as post does not have any media attached to it or its groups or options
- update endpoint GET **/api/posts/:postid** to respond with error whenever under creation post is called + tests
- update endpoint GET **/api/posts** to respond only with created posts + tests
- update endpoint PUT **/api/votes/:optionid to prevent any vote on under creation post
- update endpoint POST **/api/posts to receive the user id from autorization header and update it in the database

### CODE REVIEW

File: src/posts/posts.service.ts [hunk 1]
```diff
@@ -35,39 +40,72 @@ export class PostsService {
...
-    );
+    // get post
+    const post = await this.postRepository.findPostById(params.postid);
+
+    // check whether post is found
+    if (!post) {
+      throw new NotFoundException(`Post with id: ${params.postid} not found`);
+    }
+
+    await this.postRepository.flagPostCreation(flagPostDto.finished, post);
```
Comment (Nilegfx) [inline src/posts/posts.service.ts]: Please ensure that the user who is trying to modify the post (flag or add groups) is eligible to do so. currently the eligibility is determined by the ownership of the post. if the post owned by the requester, then pass. if not, return `403`

File: src/posts/entities/post.repository.ts [hunk 2]
```diff
@@ -37,55 +42,48 @@ export class PostRepository extends Repository<Post> {
...
   }
   /**
    * flagPostCreation
    */
-  public async flagPostCreation(flag: boolean, postid: string): Promise<void> {
-    const post = await this.findOne({ where: { uuid: postid } });
-    if (!post) throw new HttpException('Post not found', HttpStatus.NOT_FOUND);
+  public async flagPostCreation(flag: boolean, post: Post): Promise<void> {
     post.created = flag;
+    post.ready = true;
```
Comment (Nilegfx) [inline src/posts/entities/post.repository.ts]: for now, I would suggest that the ready flag follows the created flag.

File: src/posts/entities/post.repository.ts [hunk 3]
```diff
@@ -37,55 +42,48 @@ export class PostRepository extends Repository<Post> {
...
-        .leftJoin('groups.options', 'options')
-        .where('posts.uuid = :uuid', { uuid: postid })
-        .getOneOrFail();
-      return post;
-    } catch (error) {
-      if (error.name === 'EntityNotFound')
-        throw new NotFoundException('post not found');
-      else throw new InternalServerErrorException();
-    }
+    const post = await this.createQueryBuilder('posts')
```
Comment (Nilegfx) [inline src/posts/entities/post.repository.ts]: Please use singular aliases here (`post`, `group`, `option`)

File: src/posts/entities/post.repository.ts [hunk 4]
```diff
@@ -37,55 +42,48 @@ export class PostRepository extends Repository<Post> {
...
+      ])
+      .leftJoin('posts.groups', 'groups')
+      .leftJoin('groups.options', 'options')
+      .where('posts.uuid = :uuid', { uuid: postid })
+      .getOne();
+
+    return post;
+  }
+
+  public async findPostById(postId: string): Promise<Post> {
```
Comment (Nilegfx) [inline src/posts/entities/post.repository.ts]: I would suggest to name the two methods as: 
1. `findPostById`
2. `findPostDetailsById`

File: src/posts/posts.service.ts [hunk 5]
```diff
@@ -35,39 +40,72 @@ export class PostsService {
...
+    // get post
+    const post = await this.postRepository.findPostById(postid);
+
+    // check whether post is found
+    if (!post) {
+      throw new NotFoundException(`Post with id: ${postid} not found`);
+    }
+
+    // Check if current user is the owner of the post
+    if (post.user_id !== userId) {
```
Comment (Nilegfx) [inline src/posts/posts.service.ts]: Please abstract this logic to shared utility which we can use in post, votes options or any other action

File: src/posts/posts.service.ts [hunk 6]
```diff
@@ -35,39 +40,72 @@ export class PostsService {
...
   ): Promise<OptionsGroups> {
     const response: OptionsGroups = { groups: [] };
+
+    // get post
+    const post = await this.postRepository.findPostById(postid);
+
+    // check whether post found
+    if (!post) {
+      throw new NotFoundException(`Post with id: ${postid} not found`);
+    }
```
Comment (Nilegfx) [inline src/posts/posts.service.ts:109]: Please ensure to check the ownership of the post before updating it.

File: src/posts/posts.service.ts [hunk 7]
```diff
@@ -35,39 +40,72 @@ export class PostsService {
...
+
     // Loop through all groups
     for (let i = 0; i < groupsCreationDto.groups.length; i++) {
       const group = groupsCreationDto.groups[i];
       const createdGroup = await this.groupRepository.createGroup(
-        postid,
+        post,
```
Comment (Nilegfx) [inline src/posts/posts.service.ts]: Why don't you pass the options here to be created with the group creation?

File: src/posts/posts.service.ts [hunk 8]
```diff
@@ -81,7 +119,9 @@ export class PostsService {
     return response;
   }
   async getAllPosts(): Promise<Posts> {
-    const currentPosts = await this.postRepository.getAllPosts();
+    let currentPosts = await this.postRepository.getAllPosts();
```
Comment (Nilegfx) [inline src/posts/posts.service.ts]: Please refactor this method as per our discussion.

File: src/posts/posts.service.ts [hunk 9]
```diff
@@ -101,8 +141,20 @@ export class PostsService {
...
     const post = await this.postRepository.getSinglePost(postId);
+
+    // check whether post is found
+    if (!post) throw new NotFoundException(`Post with id: ${postId} not found`);
+
+    // don't return post if post.created = false
+    if (!post.created) {
+      throw new HttpException(
+        `Post with id: ${postId} still under creation...`,
+        423,
```
Comment (Nilegfx) [inline src/posts/posts.service.ts]: Please use either: 
1. extend `HttpStatus` from nest
2. extend nest expectations and introduce `LockedExpection`
behind the scene, please create `PickifyHttpStatus` enum (extending `HttpStatus` from `@nestjs/commons` or at least create our own Enum

File: src/votes/votes.service.spec.ts [hunk 10]
```diff
@@ -1,41 +1,268 @@
...
 import { Test, TestingModule } from '@nestjs/testing';
+import { Option } from '../posts/entities/option.entity';
+import { OptionRepository } from '../posts/entities/option.repository';
 import { VoteRepository } from './entities/votes.repository';
 import { VotesService } from './votes.service';
 
-const mockRepository = {
-  addVote: jest.fn(() => Promise.resolve(['options'])),
-};
+// const mockOptionDB = {
```
Comment (Nilegfx) [inline src/votes/votes.service.spec.ts]: please delete this commented code

**Review by Nilegfx — CHANGES_REQUESTED** [review]

File: src/posts/posts.service.ts (hunk 1 above)
Comment (Sahl-A) [reply src/posts/posts.service.ts]: done

File: src/posts/entities/post.repository.ts (hunk 2 above)
Comment (Sahl-A) [reply src/posts/entities/post.repository.ts]: done

File: src/posts/entities/post.repository.ts (hunk 3 above)
Comment (Sahl-A) [reply src/posts/entities/post.repository.ts]: done

File: src/posts/entities/post.repository.ts (hunk 4 above)
Comment (Sahl-A) [reply src/posts/entities/post.repository.ts]: done

File: src/posts/posts.service.ts (hunk 5 above)
Comment (Sahl-A) [reply src/posts/posts.service.ts]: done

File: src/posts/posts.service.ts (hunk 6 above)
Comment (Sahl-A) [reply src/posts/posts.service.ts:109]: done

File: src/posts/posts.service.ts (hunk 8 above)
Comment (Sahl-A) [reply src/posts/posts.service.ts]: done

File: src/posts/posts.service.ts (hunk 9 above)
Comment (Sahl-A) [reply src/posts/posts.service.ts]: Both changes are added:
1- extended `HttpStatus` using `PickifyHttpStatus`
2- Added `LockedException`

File: src/votes/votes.service.spec.ts (hunk 10 above)
Comment (Sahl-A) [reply src/votes/votes.service.spec.ts]: done

File: src/shared/enums/pickifyHttpStatus.enum.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
+import { HttpStatus } from '@nestjs/common';
+
+export const pickifyHttpStatus = {
+  ...HttpStatus,
```
Comment (Nilegfx) [inline src/shared/enums/pickifyHttpStatus.enum.ts:4]: 👍

File: src/posts/posts.service.ts (hunk 7 above)
Comment (Nilegfx) [reply src/posts/posts.service.ts]: @Sahl-A did you check this suggestion?

**Review by Nilegfx — DISMISSED** [review]
very good effort and achievements, Sahl. thank you so much for being dedicated and committed to push the project forward even we suffer resource shortage.

File: src/posts/posts.service.ts (hunk 7 above)
Comment (Sahl-A) [reply src/posts/posts.service.ts]: yes working on it.

Comment (Sahl-A) [reply src/posts/posts.service.ts]: I couldn't create the options with the group creation. However, I refactored the code to create bulk options to simplify the logic and to improve the performance by making 1 database call for multiple options.

**Review by Nilegfx — APPROVED** [review]
LGTM

**Pull request merged**